ggml-cuda.cu

```diff
@@ -81,12 +81,29 @@
 #if defined(GGML_USE_HIPBLAS)
 #define __CUDA_ARCH__ 1300
 
+#ifndef __has_builtin
+    #define __has_builtin(x) 0
+#endif
+
 typedef int8_t int8x4_t __attribute__((ext_vector_type(4)));
 static __device__ __forceinline__ int __vsubss4(const int a, const int b) {
     const int8x4_t va = reinterpret_cast<const int8x4_t&>(a);
     const int8x4_t vb = reinterpret_cast<const int8x4_t&>(b);
+#if __has_builtin(__builtin_elementwise_sub_sat)
     const int8x4_t c = __builtin_elementwise_sub_sat(va, vb);
     return reinterpret_cast<const int&>(c);
+#else
+    int8x4_t c;
+    int16_t tmp;
+#pragma unroll
+    for (int i = 0; i < 4; i++) {
+        tmp = va[i] - vb[i];
+        if(tmp > std::numeric_limits<int8_t>::max()) tmp = std::numeric_limits<int8_t>::max();
+        if(tmp < std::numeric_limits<int8_t>::min()) tmp = std::numeric_limits<int8_t>::min();
```

Comment:
If I understand this code correctly it clips the `tmp` value to be within the legal limits of `int8_t`. In that case I think something like the following would be faster:

```suggestion
        const int smaller_min = tmp < std::numeric_limits<int8_t>::min();
        const int bigger_max = tmp > std::numeric_limits<int8_t>::max();
        tmp = smaller_min * std::numeric_limits<int8_t>::min() + larger_max * std::numeric_limits<int8_t>::max()
                + !(smaller_min | larger_max) * tmp;
```

The reason I think it would be faster is that conditional statements are very slow on GPUs. Please note that I did not test this implementation and that you may need to test `int` vs. `int16_t` and `|` vs `||` for optimal performance.

Reply:
The idea behind writing it that way is for the compiler to recognize that it's been asked to do saturating subtraction and not emit any conditional instructions at all.
The rocm clangs available on compiler explorer emit `@llvm.ssub.sat.v4i8`: https://godbolt.org/z/azxe38hMs